### PR TITLE
feat(platform): MLOps training platform on EKS — Flyte + MLflow training loop (#14)

### LINKED ISSUE #14: TODO: Implement Training Loop (Optimizer, Scheduler, DataLoader)
## Overview

Once loss functions and a dataset are in place, we will need the basic training infrastructure to run learning. Listing here as a reminder of what is needed.

## Items to consider

- [ ] DataLoader for sequential frame data (depends on dataset choice)
- [ ] Optimizer selection (e.g., AdamW)
- [ ] LR Scheduler
- [ ] Basic training loop: forward → loss → backward → step
- [ ] Gradient clipping
- [ ] Logging (TensorBoard or WandB)
- [ ] Checkpointing (save/resume)
- [ ] Mixed precision (AMP) support

## Notes

The forward pass is implemented and tested. This issue tracks the remaining infrastructure needed to actually train the model. Happy to discuss priorities or implementation approach — would appreciate any guidance from the team on preferred tooling or conventions.

## Dependencies

- Dataset selection (#16)
- Loss functions (#13, #15)

### LINKED ISSUE #30: Discussion: Training-scale frame loading strategy for NVIDIA PhysicalAI dataset
## Context

In #19, @bharatwrrr added the NVIDIA PhysicalAI dataset parser. During review, we identified that the current `load_camera_frame()` implementation reads the full video file into memory on every `__getitem__` call (via `video_path.read_bytes()` + `SeekVideoReader`). This is workable for forward-pass smoke tests but would be prohibitively slow and memory-intensive for actual training loops.

@bharatwrrr addressed metadata caching (egomotion DataFrames, camera timestamps) but left the video read path as-is due to TorchCodec serialization issues with multi-worker DataLoaders. We agreed to open a follow-up issue to explore the right approach.

## Problem

For a typical 20s clip at 30 FPS with 7 cameras:
- ~600 frames per camera × 7 cameras = 4,200 frames per clip
- But only ~72 valid sample points per clip (10 Hz egomotion with 64-step history/future windows)
- Current design: reads entire video bytes, constructs a SeekVideoReader, decodes 1 frame, then discards everything — repeated per camera, per sample

With `DataLoader(num_workers=4)`, this means multiple full video files resident in memory simultaneously.

## Options discussed in #19

1. **On-the-fly TorchCodec decoding** — Replace `SeekVideoReader` with TorchCodec's seek-based decoder. Faster per-frame access without reading full bytes. However, TorchCodec objects may not serialize cleanly across DataLoader workers.

2. **TorchCodec with cached frame-index mapping** — Pre-compute which frame indices correspond to valid sample points. At training time, only decode those specific frames. Avoids keyframe penalty for distant seeks.

3. **Pre-extracted aligned-frame cache** — An offline preprocessing step that extracts only the frames aligned to valid `(clip_uuid, sample_idx, camera_name)` tuples, saving them as individual images (JPEG/PNG) or a memory-mapped tensor shard. A manifest maps `(clip_uuid, sample_idx, camera_name)` → file path.

## Questions to consider

- What's the right balance between preprocessing complexity and training-time I/O speed?
- Should we support both "raw video" and "pre-extracted" modes, so contributors can run smoke tests without preprocessing but training uses the fast path?
- Is there a preferred on-disk format for cached frames (individual JPEGs, tar archives, HDF5, memory-mapped tensors)?
- How should we handle the case where a contributor only downloads a subset of clips — should preprocessing be incremental?

## Related

- #19 (original implementation)
- #16 (Dataset Selection and DataLoader — broader scope)
- The egomotion 100Hz→10Hz downsampling (`iloc[::10]`) also warrants discussion: timestamp-based resampling with gap detection may be more robust than row-striding, but this could be a separate concern.

Would appreciate thoughts from @bharatwrrr and the team on which direction makes sense given the project's current stage.

### PR DESCRIPTION
**Pull request by riita10069**
## Summary

Implements the **training infrastructure** tracked in #14 (Training Loop) as a
production **MLOps platform on AWS EKS**, orchestrated with **Flyte** and tracked
with **MLflow** — taking AutoE2E from "forward pass on random tensors" to a real,
reproducible **data → train → evaluate** pipeline running on GPUs in the cloud.

Closes #14. Related: #16 (dataset/DataLoader), #13/#15 (losses), #30 (frame loading).

---

## What #14 asked for vs. what this delivers

| #14 checklist | Delivered |
|---------------|-----------|
| DataLoader for sequential frame data | `data_processing` pre-extracts aligned frames → WebDataset shards; `make_pre_extracted_loader` streams them (no video decode at train time, fixes #30) |
| Optimizer (AdamW) | ✅ AdamW in `train_il` |
| LR scheduler | ✅ cosine (recorded in config/metadata) |
| Basic loop: forward → loss → backward → step | ✅ real `AutoE2E` + `TrajectoryImitationLoss` |
| Gradient clipping | ✅ `grad_clip` arg |
| Logging (TB/WandB) | ✅ **MLflow** (params, metrics, artifacts, model registry) |
| Checkpointing | ✅ checkpoint + full metadata as Flyte/​S3 artifacts |
| Mixed precision (AMP) | ✅ bf16 autocast + GradScaler |

Plus the surrounding production concerns #14 implies: dataset ingestion, evaluation,
offline-RL refinement, experiment tracking, secrets, and full IaC.

---

## Architecture

```
data_ingest ─▶ data_processing ─▶ train_il ─▶ evaluate_il_policy   (MLflow: imitation-learning)
 (HF download)  (frame extract,      (AutoE2E IL)       │
                 WebDataset shards)                     ▼
                                          train_offline_rl ─▶ evaluate_rl_policy  (MLflow: offline-rl)
                                              (IQL refine)
```

- **Orchestration**: Flyte on EKS **Auto Mode** (pure — no managed node groups).
- **Each task has its own Docker image** (`data-prep`, `training`, `eval`, `offline-rl`),
  built via in-VPC CodeBuild and pushed to ECR.
- **Data passing**: FlyteDirectory/FlyteFile over S3.
- **GPU**: `g6e.4xlarge` (L40S) via a Karpenter NodePool pinned to an ODCR-backed AZ.
- **UI exposure**: CloudFront → VPC Origin → internal ALB/NLB (LBs are not
  internet-facing). Flyte Console is Cognito-protected via Lambda@Edge; MLflow is open.
- **Tracking/registry**: MLflow (RDS Postgres backend, S3 artifacts), model registry
  `auto-e2e-driving-policy`.

### Key design decisions

- **EKS Auto Mode (pure)**: built-in CNI / Karpenter / LB controller / Bottlerocket GPU
  drivers → minimal ops. Mixing Auto Mode + managed node groups causes CNI conflicts,
  so we commit to Auto Mode only.
- **CARLA dropped → Offline RL (IQL)**: CARLA needs Vulkan, which Bottlerocket lacks and
  is incompatible with the Auto Mode GPU stack. Offline RL refines the IL policy from
  recorded data, no simulator required.
- **MLflow logging only in the evaluation task**: a single point of truth means each
  MLflow run already bundles data params, model params, training hyperparams, the loss
  curve, eval metrics, `config.yaml`, and the checkpoint — fully reproducible, no
  cross-referencing a separate "training run".
- **Pre-extracted WebDataset shards (#30)**: ingestion downloads raw data; a separate
  `data_processing` step extracts only the frames aligned to valid sample points and
  packs them into tar shards. Training does sequential tar reads at disk bandwidth — no
  per-`__getitem__` video decode, worker-safe.
- **Dual-mode datasets**: the same `L2DDataset` / `NvidiaAVDataset` work locally (HF
  download) and in Flyte (pre-extracted shards), selected by an import/loader shim.
- **All datasets passed, one selected**: train/eval tasks receive every dataset's
  (separately-packed) shards and select by `dataset` arg. True multi-dataset training of
  a single model is deferred to #77 (needs runtime-dynamic `num_views` in BEV fusion).
- **Flyte S3 auth via static access key**: the stow/minio-go client (AWS SDK v1) doesn't
  support IRSA/Pod Identity, so propeller/admin storage use an access key from a
  Terraform-managed IAM user (configmap patched idempotently post-Helm).
- **Secrets**: HF token is injected into tasks via a Flyte `Secret` (K8s Secret), never a
  workflow input — so it is not visible in the Flyte/MLflow UI. No credentials in git.

---

## Repo layout (new)

```
platform/
├── infra/                      Terraform: VPC, EKS Auto Mode, RDS, ECR, S3,
│   ├── modules/{vpc,eks,rds,storage,ecr,mlflow,flyte,kueue,training-operator,
│   │            codebuild,cloudfront,auth-edge}/
│   └── environments/dev/       tfvars (+ gitignored secrets.auto.tfvars)
├── pipelines/workflows.py      Flyte-native tasks + workflows (the training loop)
├── docker/{data-prep,training,eval,offline-rl}/Dockerfile
├── helm-values/                flyte-core + mlflow values
├── k8s/                        Kueue, NodePool, namespaces, ingress
├── GETTING_STARTED.md
├── flyte_workflow_parameters.md   per-task reference
├── HowToUseFlyte.md               persona-based Flyte UI guide
└── HowToUseMLflow.md              persona-based MLflow UI guide
```

---

## Tasks (each in its own image)

| Task | Image | Role |
|------|-------|------|
| `data_ingest` | data-prep | HF download (lerobot for L2D, physical_ai_av for NVIDIA) |
| `data_processing` | data-prep | frame pre-extraction → WebDataset shards (#30) |
| `train_il` | training (GPU) | AutoE2E + TrajectoryImitationLoss, AdamW, cosine, AMP, grad-clip |
| `train_offline_rl` | offline-rl (GPU) | IQL refinement of the IL checkpoint |
| `evaluate_il_policy` | eval (GPU) | open-loop ADE/FDE → MLflow `imitation-learning` |
| `evaluate_rl_policy` | eval (GPU) | open-loop ADE/FDE → MLflow `offline-rl` |

Workflows: `wf_data_ingest`, `wf_data_processing`, `wf_train_il`,
`wf_train_offline_rl`, and the all-in-one **`wf_full_pipeline`**.

---

## Verification

- `wf_full_pipeline` runs **end-to-end to SUCCEEDED** on real data for both
  **L2D (7 cameras)** and **NVIDIA PhysicalAI-AV (8 cameras)**.
- Real metrics land in MLflow (`imitation-learning`, `offline-rl`); checkpoints are
  registered under `auto-e2e-driving-policy`.
- `terraform plan` = **No changes** (all manual fixes folded back into IaC; configs
  survive pod restarts).

## Notes / follow-ups

- Multi-dataset training of a **single** model: #77 (runtime-dynamic `num_views` BEV).
- Epoch counts default low (3) for fast smoke validation — they are hyperparameters.
- IaC is account-agnostic (no hardcoded account IDs; secrets in gitignored tfvars).

## How to use

See `platform/GETTING_STARTED.md`, `platform/HowToUseFlyte.md`, and
`platform/HowToUseMLflow.md`. TL;DR: launch `wf_full_pipeline` in the Flyte Console,
watch it in Flyte, read the metrics in MLflow.

### CODE REVIEW

**Comment by riita10069** [comment]
You can Manage and Visualize and Trigger Workflows on Flyte UI
image

You can launch training or model evaluation or something tasks on Flyte UI. super easy.
image



pipelines are basically launched from the Flyte Console (Web UI).
Select wf_full_pipeline from Workflows in the left menu, and click "Launch Workflow" to display a form.
Choose backbone and fusion_mode from the dropdown menus, enter numerical values for epochs and lr, then Launch.
With just this, the entire end-to-end process from data ingestion through training, evaluation, to Model Registry registration runs seamlessly.

**Comment by riita10069** [comment]
You can compare the result of eval on MLFlow UI.
image

image

Everything is working on my aws account and all modules are defined by Terraform.

**Comment by riita10069** [comment]
For example, when you want to investigate "which has better accuracy between concat fusion and cross_attn fusion in SwinV2-Tiny," all you need to do is the following:

1. Open the `imitation-learning` Experiment
2. Enter `` params.`model/backbone` = "swin_v2_tiny" `` in the search bar
3. Click the `eval/ade` column to sort in ascending order

This allows you to view the ADE for each fusion_mode while keeping the backbone fixed.
You can see the differences at a glance, such as "concat is 1.82m, cross_attn is 1.64m, BEV is 1.51m."

Next, when you check two Runs and press the Compare button, the parameter differences are highlighted, and the `train/loss` epoch curves are displayed overlaid.
You can visually grasp trade-offs such as "cross_attn has lower loss than concat from epoch 2 onward" and "however, training time is 1.3 times longer."

**Comment by riita10069** [comment]
@m-zain-khawaja  How do you think this platform

**Comment by apgeorg** [comment]
immense @riita10069!

**Review by m-zain-khawaja — APPROVED** [review]
Approved - thanks @riita10069 
Excellent work! Looks great 👍👏
